File: Classes/CustomUITableView.m
//
//  CustomUITableView.m
//  iSub
//
//  Created by Ben Baron on 4/27/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "CustomUITableView.h"
#import "CustomUITableViewCell.h"
#import "CellOverlay.h"

#define ISMSHorizSwipeDragMin 3
#define ISMSVertSwipeDragMax 80

#define ISMSCellEnableDelay 1.0
#define ISMSTapAndHoldDelay 0.25

@implementation CustomUITableView

@synthesize lastDeleteToggle, startTouchPosition, hasSwiped, cellShowingOverlay, tapAndHoldCell;

#pragma mark Lifecycle

- (void)setup
{
	lastDeleteToggle = [NSDate date];
    self.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
    
    if (IS_IOS7())
    {
        self.sectionIndexBackgroundColor = [UIColor clearColor];        
    }
}

- (id)initWithFrame:(CGRect)frame
{
    if ((self = [super initWithFrame:frame])) 
	{		
		[self setup];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
	if ((self = [super initWithCoder:aDecoder]))
	{
		[self setup];
	}
	return self;
}

#pragma mark Touch gestures interception

- (void)hideAllOverlays:(UITableViewCell *)cellToSkip
{
	// Remove any visible buttons
	for (UITableViewCell *cell in self.visibleCells) 
	{
		if (cell != cellToSkip)
		{
			if ([cell isKindOfClass:[CustomUITableViewCell class]])
			{
				[(CustomUITableViewCell *)cell hideOverlay];
			}
		}
	}
}

- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{	
	// Don't try anything when the tableview is moving
	// and do not catch as a swipe if touch is far right (potential index control)
	if (!self.decelerating && point.x < 290)
	{
		// Find the cell
		UITableViewCell *cell = [self cellForRowAtIndexPath:[self indexPathForRowAtPoint:point]];
		
		// Handle multi delete touching
		if (self.editing &&
			point.x < 40. && [[NSDate date] timeIntervalSinceDate:self.lastDeleteToggle] > 0.25)
		{
			self.lastDeleteToggle = [NSDate date];
			if ([cell isKindOfClass:[CustomUITableViewCell class]])
			{
				[(CustomUITableViewCell *)cell toggleDelete];
			}
		}
		
		// Remove overlays
		if (!self.hasSwiped)
		{			
			[self hideAllOverlays:cell];
			
			if ([cell isKindOfClass:[CustomUITableViewCell class]])
			{
				if ([(CustomUITableViewCell *)cell isOverlayShowing])
					[cell performSelector:@selector(hideOverlay) withObject:nil afterDelay:1.0];
			}
			
		}
	}
	//return nil;
	return [super hitTest:point withEvent:event];
}

- (BOOL)touchesShouldCancelInContentView:(UIView *)view 
{
	return YES;
}

- (BOOL)touchesShouldBegin:(NSSet *)touches withEvent:(UIEvent *)event inContentView:(UIView *)view 
{
	return YES;
}

#pragma mark Touch gestures for custom cell view

- (void)disableCellsTemporarily
{
	self.allowsSelection = NO;
	[self performSelector:@selector(enableCells) withObject:nil afterDelay:ISMSCellEnableDelay];
}

- (void)enableCells
{
	self.allowsSelection = YES;
}

- (BOOL)isTouchHorizontal:(UITouch *)touch 
{
    CGPoint currentTouchPosition = [touch locationInView:self];
    CGFloat xMovement = fabsf(startTouchPosition.x - currentTouchPosition.x);
    CGFloat yMovement = fabsf(startTouchPosition.y - currentTouchPosition.y);
    NSLog(@"xMovement: %f  yMovement: %f", xMovement, yMovement);
    
	return xMovement > yMovement;
}

- (void)lookForSwipeGestureInTouches:(NSSet *)touches withEvent:(UIEvent *)event 
{
    CGPoint currentTouchPosition = [[touches anyObject] locationInView:self];
	UITableViewCell *cell = [self cellForRowAtIndexPath: [self indexPathForRowAtPoint:currentTouchPosition]];
	
	if (!self.hasSwiped) 
	{
		// Check if this is a full swipe
		if (fabsf(self.startTouchPosition.x - currentTouchPosition.x) >= ISMSHorizSwipeDragMin 
			&& fabsf(self.startTouchPosition.y - currentTouchPosition.y) <= ISMSVertSwipeDragMax)
		{			
			self.hasSwiped = YES;
			self.scrollEnabled = NO;			
			
			// Temporarily disable the cells so we don't get accidental selections
			[self disableCellsTemporarily];
			
			// Hide any open overlays
			[self hideAllOverlays:nil];
			
			// Detect the direction
			if (self.startTouchPosition.x < currentTouchPosition.x) 
			{
				// Right swipe
				if (settingsS.isSwipeEnabled && !IS_IPAD())
				{
					if ([cell isKindOfClass:[CustomUITableViewCell class]])
					{
						[(CustomUITableViewCell *)cell showOverlay];
					}
					self.cellShowingOverlay = cell;
				}
			} 
			else 
			{
				// Left Swipe
				if ([cell isKindOfClass:[CustomUITableViewCell class]])
				{
					[(CustomUITableViewCell *)cell scrollLabels];
				}
			}
		} 
		else 
		{
			// Process a non-swipe event.
		}
	}
}

- (void)tapAndHoldFired
{
    self.hasSwiped = YES;
	if ([self.tapAndHoldCell isKindOfClass:[CustomUITableViewCell class]])
	{
		[(CustomUITableViewCell *)self.tapAndHoldCell showOverlay];
	}
	self.cellShowingOverlay = self.tapAndHoldCell;
}

- (void)cancelTapAndHold
{
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(tapAndHoldFired) object:nil];
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event 
{	
	self.allowsSelection = NO;
	self.scrollEnabled = YES;
		
	// Handle swipe
    self.startTouchPosition = [[touches anyObject] locationInView:self];
	self.hasSwiped = NO;
	self.cellShowingOverlay = nil;

	// Handle tap and hold
	if (settingsS.isTapAndHoldEnabled)
	{
		self.tapAndHoldCell = [self cellForRowAtIndexPath: [self indexPathForRowAtPoint:self.startTouchPosition]];
		[self performSelector:@selector(tapAndHoldFired) withObject:nil afterDelay:ISMSTapAndHoldDelay];
	}
	
	[super touchesBegan:touches withEvent:event];
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event 
{	
	// Cancel the tap and hold if user moves finger
	[self cancelTapAndHold];
	
	// Check for swipe
	if ([self isTouchHorizontal:[touches anyObject]]) 
	{
		[self lookForSwipeGestureInTouches:(NSSet *)touches withEvent:(UIEvent *)event];
	} 
	
	[super touchesMoved:touches withEvent:event];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event 
{
	self.allowsSelection = YES;
	self.scrollEnabled = YES;
	
	[self cancelTapAndHold];

	if (self.hasSwiped)
	{
		// Enable the buttons if the overlay is showing
		if ([self.cellShowingOverlay isKindOfClass:[CustomUITableViewCell class]])
		{
			[[(CustomUITableViewCell *)self.cellShowingOverlay overlayView] enableButtons];
		}
	}
	else
	{
		// Select the cell if this was a touch not a swipe or tap and hold
		CGPoint currentTouchPosition = [[touches anyObject] locationInView:self];
		if ((self.editing && currentTouchPosition.x > 40.) || !self.editing)
		{
			UITableViewCell *cell = [self cellForRowAtIndexPath: [self indexPathForRowAtPoint:currentTouchPosition]];
			[self selectRowAtIndexPath:[self indexPathForCell:cell] animated:NO scrollPosition:UITableViewScrollPositionNone];
			[self.delegate tableView:self didSelectRowAtIndexPath:[self indexPathForCell:cell]];
		}
	}
	self.hasSwiped = NO;
		
	[super touchesEnded:touches withEvent:event];
}

- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self cancelTapAndHold];
	
	self.allowsSelection = YES;
	self.scrollEnabled = YES;
	self.hasSwiped = NO;
	
	if ([self.cellShowingOverlay isKindOfClass:[CustomUITableViewCell class]])
	{
		[[(CustomUITableViewCell *)self.cellShowingOverlay overlayView] enableButtons];
	}
		
	[super touchesCancelled:touches withEvent:event];
}


@end

File: Classes/CustomUITableView.swift
//
//  CustomUITableView.swift
//  iSub
//
//  Created by Benjamin Baron on 12/20/14.
//  Copyright (c) 2014 Ben Baron. All rights reserved.
//

import Foundation
import UIKit

public class CustomUITableView: UITableView {
    
    let HorizSwipeDragMin: Double = 3.0
    let VertSwipeDragMax: Double = 80.0
    
    let CellEnableDelay: NSTimeInterval = 1.0
    let TapAndHoldDelay: NSTimeInterval = 0.25
    
    let _settings = SavedSettings.sharedInstance()

    var _startTouchPosition: CGPoint?
    var _swiped: Bool = false
    var _cellShowingOverlay: UITableViewCell?
    var _tapAndHoldCell: UITableViewCell?
    var _lastDeleteToggle: NSDate = NSDate()
    
    var _tapAndHoldTimer: NSTimer?
    
    private func _setup() {
        self.separatorStyle = UITableViewCellSeparatorStyle.SingleLine
        self.sectionIndexBackgroundColor = UIColor.clearColor()
    }
    
    public override init(frame: CGRect, style: UITableViewStyle) {
        super.init(frame: frame, style: style)
        self._setup()
    }
    
    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self._setup()
    }
    
    // MARK: - Touch Gestures Interception -
    
    private func _hideAllOverlays(cellToSkip: UITableViewCell?) {
        for cell in self.visibleCells as [UITableViewCell] {
            if let customCell: ItemUITableViewCell = cell as? ItemUITableViewCell {
                if customCell != cellToSkip {
                    customCell.hideOverlay()
                }
            }
        }
    }
    
    public override func hitTest(point: CGPoint, withEvent event: UIEvent?) -> UIView? {
        // Don't try anything when the tableview is moving and do not catch as a swipe if touch is far right (potential index control)
        if !self.decelerating && point.x < 290 {
            
            // Find the cell
            if let indexPath = self.indexPathForRowAtPoint(point) {
                
                if let cell = self.cellForRowAtIndexPath(indexPath) {
                    
                    // TODO: Move multi delete touch handling to ItemUITableViewCell
                    
                    // Handle multi delete touching
                    if self.editing && point.x < 40.0 && NSDate().timeIntervalSinceDate(self._lastDeleteToggle) > 0.25 {
                        self._lastDeleteToggle = NSDate()
                        if let customCell: ItemUITableViewCell = cell as? ItemUITableViewCell {
                            customCell.toggleDelete()
                        }
                    }
                    
                    // Remove overlays
                    if !self._swiped {
                        self._hideAllOverlays(cell)
                        
                        if let customCell: ItemUITableViewCell = cell as? ItemUITableViewCell {
                            if customCell.overlayShowing {
                                EX2Dispatch.runInMainThreadAfterDelay(1.0, block: {
                                    customCell.hideOverlay()
                                })
                            }
                        }
                    }
                }
            }
        }
        
        return super.hitTest(point, withEvent: event)
    }
    
    public override func touchesShouldCancelInContentView(view: UIView) -> Bool {
        return true
    }
    
    public override func touchesShouldBegin(touches: Set<UITouch>, withEvent event: UIEvent?, inContentView view: UIView) -> Bool {
        return true
    }
    
    // MARK: - Touch gestures for custom cell view -
    
    private func _disableCellsTemporarily() {
        self.allowsSelection = false
        EX2Dispatch.runInMainThreadAfterDelay(CellEnableDelay, block: {
            self.allowsSelection = true
        })
    }
    
    private func _isTouchHorizontal(touch: UITouch) -> Bool
    {
        let currentTouchPosition: CGPoint = touch.locationInView(self)
        if let startTouchPosition = self._startTouchPosition {
            let xMovement = fabs(Double(startTouchPosition.x) - Double(currentTouchPosition.x))
            let yMovement = fabs(Double(startTouchPosition.y) - Double(currentTouchPosition.y))
            
            return xMovement > yMovement
        }
        
        return false
    }
    
    private func _lookForSwipeGestureInTouches(touches: Set<UITouch>, event: UIEvent?)
    {
        var cell: UITableViewCell? = nil
        
        let currentTouchPosition: CGPoint? = touches.first?.locationInView(self)
        if currentTouchPosition != nil {
            let indexPath: NSIndexPath? = self.indexPathForRowAtPoint(currentTouchPosition!)
            if indexPath != nil {
                cell = self.cellForRowAtIndexPath(indexPath!)
            }
        }
        
        if (!self._swiped && cell != nil)
        {
            // Check if this is a full swipe
            if let startTouchPosition: CGPoint = self._startTouchPosition {
                
                let startTouchX = Double(startTouchPosition.x)
                let startTouchY = Double(startTouchPosition.y)
                let currentTouchX = Double(currentTouchPosition!.x)
                let currentTouchY = Double(currentTouchPosition!.y)
                
                let xDiff = fabs(startTouchX - currentTouchX)
                let yDiff = fabs(startTouchY - currentTouchY)
                
                if xDiff >= HorizSwipeDragMin && yDiff <= VertSwipeDragMax {
                        self._swiped = true
                        self.scrollEnabled = false
                        
                        // Temporarily disable the cells so we don't get accidental selections
                        self._disableCellsTemporarily()
                        
                        // Hide any open overlays
                        self._hideAllOverlays(nil)
                        
                        // Detect the direction
                        if startTouchX < currentTouchX {
                            // Right swipe
                            if _settings.isSwipeEnabled && !IS_IPAD() {
                                if let customCell = cell as? ItemUITableViewCell {
                                    customCell.showOverlay()
                                }
                                
                                self._cellShowingOverlay = cell
                            }
                        } else {
                            // Left Swipe
                            if let customCell = cell as? ItemUITableViewCell {
                                customCell.scrollLabels()
                            }
                        }
                } else {
                    // Process a non-swipe event.
                }
            }
        }
    }
    
    public func _tapAndHoldFired() {
        self._swiped = true
        if let customCell: ItemUITableViewCell = self._tapAndHoldCell as? ItemUITableViewCell {
            customCell.showOverlay()
        }
        self._cellShowingOverlay = self._tapAndHoldCell
    }
    
    private func _cancelTapAndHold() {
        _tapAndHoldTimer?.invalidate();
        _tapAndHoldTimer = nil
    }
    
    public override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
    
        self.allowsSelection = false
        self.scrollEnabled = true
    
        // Handle swipe
        if let touch = touches.first {
            self._startTouchPosition = touch.locationInView(self)
            self._swiped = false
            self._cellShowingOverlay = nil
            
            // Handle tap and hold
            if (_settings.isTapAndHoldEnabled) {
                let indexPath = self.indexPathForRowAtPoint(self._startTouchPosition!)
                if let indexPath = indexPath {
                    self._tapAndHoldCell = self.cellForRowAtIndexPath(indexPath)
                    
                    _tapAndHoldTimer = NSTimer.scheduledTimerWithTimeInterval(TapAndHoldDelay, target: self, selector: "_tapAndHoldFired", userInfo: nil, repeats: false);
                }
            }
        }
        
        super.touchesBegan(touches, withEvent: event)
    }
    
    public override func touchesMoved(touches: Set<UITouch>, withEvent event: UIEvent?) {
        // Cancel the tap and hold if user moves finger
        self._cancelTapAndHold()
        
        // Check for swipe
        if let touch = touches.first {
            if self._isTouchHorizontal(touch) {
                self._lookForSwipeGestureInTouches(touches, event: event)
            }
        }
        
        super.touchesMoved(touches, withEvent: event)
    }

    public override func touchesEnded(touches: Set<UITouch>, withEvent event: UIEvent?) {
        self.allowsSelection = true
        self.scrollEnabled = true
        
        self._cancelTapAndHold()
        
        if self._swiped {
            // Enable the buttons if the overlay is showing
            if let customCell: ItemUITableViewCell = self._cellShowingOverlay as? ItemUITableViewCell {
                customCell.overlayView?.enableButtons();
            }
        } else if let touch = touches.first {
            // Select the cell if this was a touch not a swipe or tap and hold
            
            let currentTouchPosition = touch.locationInView(self)
            if (self.editing && Float(currentTouchPosition.x) > 40.0) || !self.editing {
                let indexPath: NSIndexPath? = self.indexPathForRowAtPoint(currentTouchPosition)
                
                if indexPath != nil {
                    self.selectRowAtIndexPath(indexPath, animated: false, scrollPosition: .None)
                    if let delegate: UITableViewDelegate = self.delegate {
                        delegate.tableView?(self, didSelectRowAtIndexPath: indexPath!)
                    }
                }
            }
        }
        self._swiped = false
        
        super.touchesEnded(touches, withEvent: event)
    }
    
    public override func touchesCancelled(touches: Set<UITouch>?, withEvent event: UIEvent?) {
        self._cancelTapAndHold()
        
        self.allowsSelection = true
        self.scrollEnabled = true
        self._swiped = false
        
        if let customCell: ItemUITableViewCell = self._cellShowingOverlay as? ItemUITableViewCell {
            customCell.overlayView?.enableButtons();
        }
        
        super.touchesCancelled(touches, withEvent: event)
    }
}
